source/fuzz/transformation_outline_selection_construct.cpp

```diff
@@ -0,0 +1,178 @@
+         "Should condition should've been checked in the IsApplicable");
+
+  const auto successor_id =
+      new_header_block->terminator()->GetSingleWordInOperand(0);
+
+  // Change |entry_block|'s terminator to |OpBranchConditional|.
+  new_header_block->terminator()->SetOpcode(SpvOpBranchConditional);
+  new_header_block->terminator()->SetInOperands(
+      {{SPV_OPERAND_TYPE_ID,
+        {fuzzerutil::MaybeGetBoolConstant(ir_context, *transformation_context,
+                                          true, true)}},
+       {SPV_OPERAND_TYPE_ID, {successor_id}},
+       {SPV_OPERAND_TYPE_ID, {successor_id}}});
+
+  // Insert OpSelectionMerge before the terminator.
+  new_header_block->terminator()->InsertBefore(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpSelectionMerge, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {message_.new_merge_block_id()}},
+          {SPV_OPERAND_TYPE_SELECTION_CONTROL,
+           {SpvSelectionControlMaskNone}}}));
+
+  // We've change the module so we must invalidate analyses.
+  ir_context->InvalidateAnalysesExceptFor(opt::IRContext::kAnalysisNone);
+}
+
+protobufs::Transformation TransformationOutlineSelectionConstruct::ToMessage()
+    const {
+  protobufs::Transformation result;
+  *result.mutable_outline_selection_construct() = message_;
+  return result;
+}
+
+bool TransformationOutlineSelectionConstruct::IsApplicableToBlockRange(
+    opt::IRContext* ir_context, uint32_t header_block_candidate_id,
+    uint32_t merge_block_candidate_id) {
+  // Check that |header_block_candidate_id| and |merge_block_candidate_id| are
+  // valid.
+  for (auto block_id : {header_block_candidate_id, merge_block_candidate_id}) {
+    const auto* label_inst = ir_context->get_def_use_mgr()->GetDef(block_id);
+    if (!label_inst || label_inst->opcode() != SpvOpLabel) {
+      return false;
+    }
+  }
+
+  const auto* header_block_candidate =
+      ir_context->cfg()->block(header_block_candidate_id);
+  const auto* merge_block_candidate =
+      ir_context->cfg()->block(merge_block_candidate_id);
+
+  // |header_block_candidate| and |merge_block_candidate| must be from the same
+  // function.
+  if (header_block_candidate->GetParent() !=
+      merge_block_candidate->GetParent()) {
+    return false;
+  }
+
+  const auto* dominator_analysis =
+      ir_context->GetDominatorAnalysis(header_block_candidate->GetParent());
+  const auto* postdominator_analysis =
+      ir_context->GetPostDominatorAnalysis(header_block_candidate->GetParent());
+
+  if (!dominator_analysis->Dominates(header_block_candidate,
+                                     merge_block_candidate) ||
+      !postdominator_analysis->Dominates(merge_block_candidate,
+                                         header_block_candidate)) {
+    return false;
+  }
+
+  // |header_block_candidate| can't be a header since we are about to make it
+  // one.
+  if (header_block_candidate->GetMergeInst()) {
```

Comment:
Also use `header_block_candidate->IsLoopHeader()` here.

Reply:
In this case, `IsLoopHeader` won't be enough since it doesn't check for `OpSelectionMerge` instruction.

Reply:
You're right.